File: src/relax/transform/tuning_api/primitives.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * \file src/relax/transform/tuning_api/primitives.cc
 * \brief Implementation of tuning APIs.
 */

#include <tvm/relax/tuning_api.h>

#include "../../../meta_schedule/utils.h"
namespace tvm {
namespace relax {

Choice::Choice(String f_transform_key, Array<ObjectRef> f_transform_args, String f_constr_key,
               Array<ObjectRef> f_constr_args) {
  ObjectPtr<ChoiceNode> n = make_object<ChoiceNode>();
  n->f_transform_key = std::move(f_transform_key);
  n->f_transform_args = std::move(f_transform_args);
  n->f_constr_key = std::move(f_constr_key);
  n->f_constr_args = std::move(f_constr_args);
  data_ = std::move(n);
}

// TODO (sunggg): Currently, it only supports an array of primitive data types.
ObjectRef ChoiceNode::AsJSON() const {
  Array<ObjectRef> json_transfrom_args, json_constr_args;
  for (ObjectRef arg : this->f_transform_args) {
    std::string json_arg = tvm::SaveJSON(arg);
    std::string b64_arg = meta_schedule::Base64Encode(json_arg);
    json_transfrom_args.push_back(String(b64_arg));
  }
  for (ObjectRef arg : this->f_constr_args) {
    std::string json_arg = tvm::SaveJSON(arg);
    std::string b64_arg = meta_schedule::Base64Encode(json_arg);
    json_constr_args.push_back(String(b64_arg));
  }
  return Array<ObjectRef>{
      this->f_transform_key,
      json_transfrom_args,
      this->f_constr_key,
      json_constr_args,
  };
}

Choice Choice::FromJSON(const ObjectRef& json) {
  // Parse `json` into `choice`
  String f_transform_key, f_constr_key;
  Array<ObjectRef> f_transform_args, f_constr_args;
  try {
    const ArrayNode* arr = json.as<ArrayNode>();
    ICHECK(arr && arr->size() == 4);
    const auto* arr0 = arr->at(0).as<StringObj>();
    const auto* arr1 = arr->at(1).as<ArrayNode>();
    const auto* arr2 = arr->at(2).as<StringObj>();
    const auto* arr3 = arr->at(3).as<ArrayNode>();
    ICHECK(arr0 && arr1 && arr2 && arr3);
    f_transform_key = GetRef<String>(arr0);
    {
      f_transform_args.reserve(arr1->size());
      for (const ObjectRef& elem : *arr1) {
        String b64_arg = Downcast<String>(elem);
        std::string json_arg = meta_schedule::Base64Decode(b64_arg);
        ObjectRef arg = LoadJSON(json_arg);
        f_transform_args.push_back(arg);
      }
    }
    f_constr_key = GetRef<String>(arr2);
    {
      f_constr_args.reserve(arr3->size());
      for (const ObjectRef& elem : *arr3) {
        String b64_arg = Downcast<String>(elem);
        std::string json_arg = meta_schedule::Base64Decode(b64_arg);
        ObjectRef arg = LoadJSON(json_arg);
        f_constr_args.push_back(arg);
      }
    }

  } catch (const tvm::Error& e) {
    LOG(FATAL)
        << "ValueError: The json entry of a choice should contain a set of two strings, but gets: "
        << json;
    throw;
  }
  return Choice(f_transform_key, f_transform_args, f_constr_key, f_constr_args);
}

Knob::Knob(String name, Map<String, Choice> choices) {
  ObjectPtr<KnobNode> n = make_object<KnobNode>();
  n->name = std::move(name);
  n->choices = std::move(choices);
  data_ = std::move(n);
}

ObjectRef KnobNode::AsJSON() const {
  Map<String, ObjectRef> json_choices;
  for (auto const& x : choices) {
    json_choices.Set(x.first, x.second->AsJSON());
  }
  return Array<ObjectRef>{
      /* 0: name    */ std::move(name),
      /* 1: choices */ std::move(json_choices),
  };
}

Knob Knob::FromJSON(const ObjectRef& json) {
  // Parse `json` into `name` and `choices`
  String name;
  Map<String, Choice> choices;
  try {
    const ArrayNode* arr = json.as<ArrayNode>();
    ICHECK(arr && arr->size() == 2);
    const auto* arr0 = arr->at(0).as<StringObj>();
    const auto* arr1 = arr->at(1).as<MapNode>();
    ICHECK(arr0 && arr1);
    name = GetRef<String>(arr0);
    for (auto const& x : GetRef<Map<String, ObjectRef>>(arr1)) {
      String decision = x.first;
      Choice choice = Choice::FromJSON(x.second);
      choices.Set(decision, choice);
    }
  } catch (const tvm::Error& e) {
    LOG(FATAL)
        << "ValueError: The json entry of a choice should contain a set of two strings, but gets: "
        << json;
    throw;
  }
  return Knob(name, choices);
}

Trace::Trace() { data_ = make_object<TraceNode>(); }

Trace::Trace(IRModule in_mod, Array<Knob> knobs, Array<String> decisions) {
  ICHECK(knobs.size() == decisions.size()) << "Size of knobs and decisions should match";
  // Deep-copy IRModule
  IRModule out_mod = meta_schedule::DeepCopyIRModule(in_mod);
  // Apply the decision history if provided
  int size = knobs.size();
  for (int i = 0; i < size; i++) {
    out_mod = knobs[i]->Apply(out_mod, decisions[i]);
  }

  ObjectPtr<TraceNode> n = make_object<TraceNode>();
  n->in_mod = std::move(in_mod);
  n->out_mod = std::move(out_mod);
  n->knobs = std::move(knobs);
  n->decisions = std::move(decisions);
  n->size = std::move(size);
  data_ = std::move(n);
}

ObjectRef TraceNode::AsJSON() const {
  ICHECK(this->Verify()) << "Trace should be valid";
  std::string json_mod = tvm::SaveJSON(this->in_mod);
  std::string b64_mod = meta_schedule::Base64Encode(json_mod);

  Array<ObjectRef> json_knobs;
  Array<ObjectRef> json_decisions;

  int size = this->size;
  json_knobs.reserve(size);
  json_decisions.reserve(size);

  for (int i = 0; i < size; i++) {
    const Knob& knob = this->knobs[i];
    const String& decision = this->decisions[i];

    json_knobs.push_back(knob->AsJSON());
    json_decisions.push_back(decision);
  }

  return Array<ObjectRef>{String(b64_mod), json_knobs, json_decisions};
}

Trace Trace::FromJSON(const ObjectRef& json) {
  // Parse `json` into `trace`
  IRModule in_mod;
  Array<Knob> knobs;
  Array<String> decisions;
  try {
    const ArrayNode* arr = json.as<ArrayNode>();
    ICHECK(arr && arr->size() == 3);
    const auto* arr0 = arr->at(0).as<StringObj>();
    const auto* arr1 = arr->at(1).as<ArrayNode>();
    const auto* arr2 = arr->at(2).as<ArrayNode>();
    ICHECK(arr0 && arr1 && arr2);

    String b64_mod = GetRef<String>(arr0);
    std::string json_mod = meta_schedule::Base64Decode(b64_mod);
    in_mod = Downcast<IRModule>(LoadJSON(json_mod));

    for (const ObjectRef& elem : *arr1) {
      knobs.push_back(Knob::FromJSON(elem));
    }

    for (const ObjectRef& elem : *arr2) {
      decisions.push_back(Downcast<String>(elem));
    }
  } catch (const tvm::Error& e) {
    LOG(FATAL)
        << "ValueError: The json entry of a choice should contain a set of two strings, but gets: "
        << json;
    throw;
  }
  return Trace(in_mod, knobs, decisions);
}

/**************** FFI ****************/
TVM_REGISTER_NODE_TYPE(ChoiceNode);
TVM_REGISTER_GLOBAL("relax.tuning_api.Choice")
    .set_body_typed([](String f_transform_key, Array<ObjectRef> f_transform_args,
                       String f_constr_key, Array<ObjectRef> f_constr_args) {
      return Choice(f_transform_key, f_transform_args, f_constr_key, f_constr_args);
    });
TVM_REGISTER_GLOBAL("relax.tuning_api.ChoiceAsJSON").set_body_method<Choice>(&ChoiceNode::AsJSON);
TVM_REGISTER_GLOBAL("relax.tuning_api.ChoiceFromJSON").set_body_typed(Choice::FromJSON);
TVM_REGISTER_GLOBAL("relax.tuning_api.ChoiceGetTransformFunc")
    .set_body_method<Choice>(&ChoiceNode::GetTransformFunc);
TVM_REGISTER_GLOBAL("relax.tuning_api.ChoiceGetConstrFunc")
    .set_body_method<Choice>(&ChoiceNode::GetConstrFunc);
TVM_REGISTER_GLOBAL("relax.tuning_api.ChoiceApplyTransformFunc")
    .set_body_method<Choice>(&ChoiceNode::ApplyTransformFunc);
TVM_REGISTER_GLOBAL("relax.tuning_api.ChoiceCheckConstr")
    .set_body_method<Choice>(&ChoiceNode::CheckConstr);

TVM_REGISTER_NODE_TYPE(KnobNode);
TVM_REGISTER_GLOBAL("relax.tuning_api.Knob")
    .set_body_typed([](String name, Map<String, Choice> choices) { return Knob(name, choices); });
TVM_REGISTER_GLOBAL("relax.tuning_api.KnobAsJSON").set_body_method<Knob>(&KnobNode::AsJSON);
TVM_REGISTER_GLOBAL("relax.tuning_api.KnobFromJSON").set_body_typed(Knob::FromJSON);
TVM_REGISTER_GLOBAL("relax.tuning_api.KnobVerify").set_body_method<Knob>(&KnobNode::Verify);
TVM_REGISTER_GLOBAL("relax.tuning_api.KnobApply").set_body_method<Knob>(&KnobNode::Apply);

TVM_REGISTER_NODE_TYPE(TraceNode);
TVM_REGISTER_GLOBAL("relax.tuning_api.Trace")
    .set_body_typed([](IRModule in_mod, Array<Knob> knobs, Array<String> decisions) {
      return Trace(in_mod, knobs, decisions);
    });
TVM_REGISTER_GLOBAL("relax.tuning_api.TraceVerify").set_body_method<Trace>(&TraceNode::Verify);
TVM_REGISTER_GLOBAL("relax.tuning_api.TraceAdd").set_body_method<Trace>(&TraceNode::Add);
TVM_REGISTER_GLOBAL("relax.tuning_api.TraceSetPerf").set_body_method<Trace>(&TraceNode::SetPerf);
TVM_REGISTER_GLOBAL("relax.tuning_api.TraceSetOutMod")
    .set_body_method<Trace>(&TraceNode::SetOutMod);

TVM_REGISTER_GLOBAL("relax.tuning_api.TraceAsJSON").set_body_method<Trace>(&TraceNode::AsJSON);
TVM_REGISTER_GLOBAL("relax.tuning_api.TraceFromJSON").set_body_typed(Trace::FromJSON);
}  // namespace relax
}  // namespace tvm

File: python/tvm/relax/transform/tuning_api/primitives.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Relax Tuning Pass API primitives"""

from typing import Callable, Union, Dict, List, Optional
import logging
import tvm
from tvm.runtime import Object
from tvm.ir.module import IRModule
from tvm.relax import Expr
from tvm.tir.schedule.trace import JSON_TYPE, _json_from_tvm
from tvm._ffi import register_object
from . import _ffi_api

logger = logging.getLogger("TuningAPI")  # pylint: disable=invalid-name


@register_object("relax.tuning_api.Choice")
class Choice(Object):
    """
    A TVM object Choice that maintains a set of transformation and constraint function keys.
    Corresponding functions should be registered as PackedFunc with these keys.
    Transformation function will be applied when constraint function returns true.
    Parameters
    ----------
    f_transform_key : Optional[str]
        Key for transformation function.
    f_transform_args : Optional[List]
        Arguments for transformation function.
    f_constr_key : Optional[str]
        Key for constraint function.
    f_constr_args : Optional[List]
        Arguments for constraint function.

    Examples
    --------
    The following code block defines a Choice.

    .. code-block:: python
        @tvm.register_func("relax.tuning_api.test.f_transform")
        def apply(mod):
            return relax.tuning_api.FoldConstant()(mod)
        @tvm.register_func("relax.tuning_api.test.f_constr")
        def constr(mod):
            return len(mod.functions) == 3
        # Define a choice to apply constant folding only when IRModule has three functions.
        choice = Choice(f_transform_key = "relax.tuning_api.test.f_transform", f_constr_key = "relax.tuning_api.test.f_constr")
    """

    def __init__(
        self,
        f_transform_key: Optional[str] = None,
        f_transform_args: Optional[List] = None,
        f_constr_key: Optional[str] = None,
        f_constr_args: Optional[List] = None,
    ):
        """Constructor
        Parameters
        ----------
        f_transform_key : Optional[str]
            Key for transformation function.

        f_tramsform_args: Optional[List]
            Arguments for transformation function.

        f_constr_key : Optional[str]
            Key for constraint function.

        f_constr_args: Optional[List]
            Arguments for constraint function.
        """

        if f_transform_key is None:
            f_transform_key = "relax.tuning_api.Choice.f_default_transform"

        if f_transform_args is None:
            f_transform_args = []

        if f_constr_key is None:
            f_constr_key = "relax.tuning_api.Choice.f_default_constr"

        if f_constr_args is None:
            f_constr_args = []

        self.__init_handle_by_constructor__(
            _ffi_api.Choice,
            f_transform_key,
            f_transform_args,
            f_constr_key,
            f_constr_args,  # type: ignore # pylint: disable=no-member
        )

    def get_transform_func(self) -> Callable:
        """Getter for f_transform
        Returns
        -------
        ret: Callable
           registered transformation function
        """
        return _ffi_api.ChoiceGetTransformFunc(self)

    def get_constr_func(self) -> Callable:
        """Getter for f_constr
        Returns
        -------
        ret: Callable
           registered constraint function
        """
        return _ffi_api.ChoiceGetConstrFunc(self)

    def apply_transform_func(self, mod: IRModule) -> IRModule:
        """Perform f_transform with its arguments
        Returns
        -------
        ret: Callable
           registered transformation function
        """
        return _ffi_api.ChoiceApplyTransformFunc(self, mod)

    def check_constr(self, mod: IRModule) -> bool:
        """Perform f_constr with its arguments
        Returns
        -------
        ret: bool
           Returns whether the IRModule satisfies the constraint or not
        """
        return _ffi_api.ChoiceCheckConstr(self, mod)

    def as_json(self) -> JSON_TYPE:
        """Serialize the trace as a JSON-style object
        Returns
        -------
        json: JSON_TYPE
            The JSON-style object
        """
        return _ffi_api.ChoiceAsJSON(self)  # type: ignore # pylint: disable=no-member

    @staticmethod
    def from_json(json_obj: JSON_TYPE) -> "Choice":
        """Create Choice from JSON obj

        Parameters
        ----------
        json_obj: JSON_TYPE
            Choice serialized with JSON

        Return
        ----------
        choice: Choice
            Deserialized choice
        """
        return _ffi_api.ChoiceFromJSON(json_obj)


@register_object("relax.tuning_api.Knob")
class Knob(Object):
    """
    A TVM object Knob that maintains a set of valid Choices.
    By using Knobs, a tuning pass can generate candidates and define the search space.
    Parameters
    ----------
    name : str
        Name of the knob.

    choices: Union[List[Choice], Dict[str, Choice]]
        A list of valid choices

    Examples
    --------
    The following code block defines a Knob.

    .. code-block:: python
        @tvm.register_func("relax.tuning_api.test.f_transform")
        def apply(mod):
            return relax.tuning_api.FoldConstant()(mod)
        choices = {"apply": Choice("relax.tuning_api.test.f_transform"), "noapply": Choice()}
        # A knob manages a set of its valid choices
        knob = Knob("MockTuningKnob", choices)
    """

    def __init__(self, name: str, choices: Union[List[Choice], Dict[str, Choice]]):
        """Constructor."""
        if isinstance(choices, list):
            choices = {str(idx): val for idx, val in enumerate(choices)}

        self.__init_handle_by_constructor__(
            _ffi_api.Knob, name, choices  # type: ignore # pylint: disable=no-member
        )

    def verify(self, decision: Union[str, int]) -> bool:
        """Verify if the decision is valid."""
        if isinstance(decision, int):
            decision = str(decision)
        return _ffi_api.KnobVerify(self, decision)

    def apply(self, mod: IRModule, decision: Union[str, int]) -> IRModule:
        """Get choice if a decision is valid."""
        if isinstance(decision, int):
            decision = str(decision)
        return _ffi_api.KnobApply(self, mod, decision)

    def as_json(self) -> JSON_TYPE:
        """Serialize the trace as a JSON-style object
        Returns
        -------
        json: JSON_TYPE
            The JSON-style object
        """
        return _ffi_api.KnobAsJSON(self)

    @staticmethod
    def from_json(json_obj: JSON_TYPE) -> "Knob":
        """Create Knob from JSON obj

        Parameters
        ----------
        json_obj: JSON_TYPE
            Knob serialized with JSON

        Return
        ----------
        knob: Knob
            Deserialized knob
        """
        return _ffi_api.KnobFromJSON(json_obj)

    def __str__(self) -> str:
        msg = f"{self.name} (# of choices: {len(self.choices)})\n"
        for name, choice in self.choices.items():
            msg += f"  - {name}: {choice}\n"
        return msg


@register_object("relax.tuning_api.Trace")
class Trace(Object):
    """
    A TVM object Trace logs the history of transformations (decisions).
    Parameters
    ----------
    in_mod : IRModule
        Input IRModule.
    knobs: Optional[List[Knob]]
        A list of knobs applied in the trace.
    decisions: Optional[List[Union[str, int]]]
        A list of decisions made for each knob

    Examples
    --------
    The following code block defines a Trace.

    .. code-block:: python

        trace = Trace(mod, [knob1, knob2, knob3], ["c1", "c0", "c3"])
        assert trace.size == 3 # Length of history.
        # 'out' contains IRModule that applies transformations in the trace.
        out: IRModule = trace.add(knob4, "c2")
        assert trace.size == 4 # Length of history.
        trace.set_perf(0.03) # Set the performance number of the trace.
    """

    def __init__(
        self,
        in_mod: IRModule,
        knobs: Optional[List[Knob]] = None,
        decisions: Optional[List[Union[str, int]]] = None,
    ):
        """Constructor."""
        knobs = knobs if knobs else list()
        decisions = (
            [str(v) if isinstance(v, int) else v for v in decisions] if decisions else list()
        )
        self.__init_handle_by_constructor__(
            _ffi_api.Trace, in_mod, knobs, decisions  # type: ignore # pylint: disable=no-member
        )

    def verify(self) -> bool:
        """Verify if current history is valid."""
        return _ffi_api.TraceVerify()

    def add(self, knob: Knob, decision: Union[str, int]) -> IRModule:
        """Add & Apply new decision (with knob)."""
        if isinstance(decision, int):
            decision = str(decision)
        return _ffi_api.TraceAdd(self, knob, decision)

    def set_perf(self, perf: float) -> None:
        """Set performance number for the trace."""
        return _ffi_api.TraceSetPerf(self, perf)

    def set_out_mod(self, mod: IRModule) -> None:
        """Set out_mod for the trace."""
        return _ffi_api.TraceSetOutMod(self, mod)

    def as_json(self, include_irmod: bool = True) -> JSON_TYPE:
        """Serialize the trace as a JSON-style object.
        Parameters
        ----------
        include_irmod: bool
            Decides whether to serialize in_mod as well.

        Returns
        -------
        json: JSON_TYPE
            The JSON-style object.
        """
        obj = _ffi_api.TraceAsJSON(self, include_irmod)
        return _json_from_tvm(obj)

    @staticmethod
    def from_json(json_obj: JSON_TYPE) -> "Trace":
        """Create Trace from JSON obj.

        Parameters
        ----------
        json_obj: JSON_TYPE
            Trace serialized with JSON.

        Return
        ----------
        trace: Trace
            Deserialized trace.
        """
        return _ffi_api.TraceFromJSON(json_obj)

    def __str__(self) -> str:
        n = len(self.knobs)
        msg = f"Trace length: {n}\n"
        for idx in range(n):
            msg += f"[{idx+1}] {self.knobs[idx].name}: {self.decisions[idx]}\n"
        return msg


def get_trace(in_: Union[Trace, IRModule, Expr]) -> Trace:
    """
    Getter for a trace wrapper.

    Parameters
    ----------
    in_: Union[Trace, IRModule, Expr]
        Input entity
    Return
    ----------
    wrapped: Trace
        Traced entity
    """
    if isinstance(in_, Trace):
        return in_
    if isinstance(in_, IRModule):
        return Trace(in_)
    if isinstance(in_, Expr):
        return Trace(tvm.IRModule.from_expr(in_))

    raise Exception(f"Invalid input type for trace: {type(in_)}")
